packages/salesforcedx-vscode-soql/test/vscode-integration/client/client.test.ts

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { expect } from 'chai';
+import * as path from 'path';
+import * as sinon from 'sinon';
+import { extensions, languages, Uri, window, workspace } from 'vscode';
+import { clearDiagnostics } from '../../../src/client/client';
+import { getMockConnection, MockConnection } from '../testUtilities';
+import * as vscode from 'vscode';
+
+const sfdxCoreExtension = vscode.extensions.getExtension(
+  'salesforce.salesforcedx-vscode-core'
+);
+const sfdxCoreExports = sfdxCoreExtension?.exports;
+const { workspaceContext } = sfdxCoreExports;
+
+async function sleep(ms: number = 0) {
+  return new Promise(resolve => {
+    setTimeout(resolve, ms);
+  });
+}
+
+function waitUntil(predicate: () => boolean) {
+  return new Promise(async resolve => {
+    let notFound = true;
+    let tries = 5;
+    while (notFound || tries-- > 0) {
+      await sleep(50);
+      notFound = !predicate();
+    }
+    resolve();
+  });
```

Comment:
Could this possibly cause flapping tests? What's the consequence of it not succeeding?

Reply:
Hopefully not. LSP error diagnostics message arrive asynchronously, and it takes a few milliseconds on the test.
See usage below (line 99):
```
    await waitUntil(() => {
      return languages.getDiagnostics(soqlFileUri).length > 0;
    });
```

Reply:
Gotcha. I don't think this is blocking btw but if this becomes a problem or you'd like to address it earlier maybe you could mock the lsp response and make it less of an integration test.

Reply:
👍  understood. Thanks.

This case is weird:

1. The client changes the file contents
1. The server gets notified, and then it's the server that sends a request to the client to run the query
1. The client runs it and responds (`client.onRequest( ... ) `)
1. If the query returned errors, we expect the server to push diagnostics back to the client

That's the flow under test here. IMHO, the main point of this test is the integration... if the test mocks it, then it'd not be testing much at all.

Unfortunately, VScode integration tests are apparently not fully real either: I tried the "serialization" issue I mentioned above and it is not caught during automated tests (may be it uses shared memory instead of serializing the objects?).